### ISSUE TITLE
Configuring the architecture of the neural net to be trained.

### ISSUE THREAD

**Issue by rajatpaliwal** [Unity-Technologies/ml-agents#2518]
Is there any way to look into the architecture of the neural net being trained and then reconfiguring the parameters according to our requirement in order to achieve better results. I know that frozen_graph_def.pb contain the information regarding the latest trained neural net, but I am unable to read it. Any suggestion would be helpful.

**Comment by chriselion** [comment]
Hi @rajatpaliwal,
I'm looking into the best way to do this; it looks like we don't have an option for it currently, but you can add
```python
self.summary_writer.add_graph(self.policy.graph, step)
```
to Trainer.write_summary(), and a visualization of the graph will be available in Tensorboard (see https://www.tensorflow.org/tensorboard/r1/graphs).

I don't think allowing editing of the graph is something that we plan to support.

**Comment by rajatpaliwal** [comment]
Hi @chriselion ,
Just confirming, so I should add this line in trainer.py inside def write_summary?

**Comment by rajatpaliwal** [comment]
I added the line "self.summary_writer.add_graph(self.policy.graph, step)" in trainer.write_summary(). But in the tensorboard i am seeing no graph definitions files were found.

**Comment by chriselion** [comment]
Not sure about that warning. I think adding 
`self.summary_writer.add_graph(self.policy.graph)`
in Trainer.save_model() might be a better place instead; that gets called if you stop training early.

**Comment by rajatpaliwal** [comment]
Thanks, that worked. Able to visualize the graph now. Do you think tensorflow graph transform tools can help in editing this neural net architecture.

**Comment by chriselion** [comment]
I'm not familiar with the tensorflow graph transform tools, and it's not something that we can provide any support for.

Backing up a bit, what are you trying to accomplish by editing the network?

**Comment by rajatpaliwal** [comment]
Its just that in order to achieve better training results I believe apart from setting hyperparameters in config files, configuring the architecture of the neural network by deciding upon the loss function being used, dropout techniques etc. can help in achieving better results.

**Comment by rajatpaliwal** [comment]
Hi @chriselion , Thanks for guiding me to this file. This was really helpful. I believe I have quite a lot of insight into the neural net architecture we are using and the way to configure it according to me.

**Comment by chriselion** [comment]
Hi @rajatpaliwal,
I logged a feature request in our internal tracker to automatically save the graph for Tensorboard.
Closing this issue for now, but please reopen (or make a new one) if you have more problems.

**Comment by rajatpaliwal** [comment]
Hi @chriselion ,
I have another small query which is how can I add a histogram depicting the weights of hidden layers after the training is over.
Any suggestion would be helpful.

**Comment by rajatpaliwal** [comment]
Thanks @chriselion  for the suggestion. Will be waiting for anymore suggestions from your side.

**Comment by chriselion** [comment]
I asked around here but nobody else has used the histograms either.

If you'd like to experiment with it and submit a pull request, we'd be happy to take the change, but otherwise I don't think it's something that we're likely to add.

**Comment by MarkTension** [comment]
> Not sure about that warning. I think adding
> `self.summary_writer.add_graph(self.policy.graph)`
> in Trainer.save_model() might be a better place instead; that gets called if you stop training early.

Any updates doing this with ml agents 0.14?

**Comment by tmendozarias** [comment]
Is self.summary_writer.add_graph(self.policy.graph) still not present in the latest release?

**Comment by tmendozarias** [comment]
> Not sure about that warning. I think adding
> `self.summary_writer.add_graph(self.policy.graph)`
> in Trainer.save_model() might be a better place instead; that gets called if you stop training early.

I Tried doing this, however I get an error saying that PPOTrainer does not have an attribute "summary_writer". I also tried with "summary_writers" but still no luck...anyone else knows how to do this in MlAgents 1.0 Release 3? Thanks!
